runtime/stdlib/contract_update_validation.go

```diff
@@ -334,18 +374,46 @@ func (validator *ContractUpdateValidator) oldTypeID(oldType *ast.NominalType) co
 	return oldImportLocation.TypeID(nil, qualifiedIdentifier)
 }
 
+func checkTypeNotRemoved(
+	validator UpdateValidator,
+	newDeclaration ast.Declaration,
+	removedTypes map[string]struct{},
+) {
+	if _, present := removedTypes[newDeclaration.DeclarationIdentifier().Identifier]; present {
+		validator.report(&UseOfRemovedTypeError{
+			Declaration: newDeclaration,
+			Range:       ast.NewUnmeteredRangeFromPositioned(newDeclaration),
+		})
+	}
+}
+
 func checkNestedDeclarations(
 	validator UpdateValidator,
 	oldDeclaration ast.Declaration,
 	newDeclaration ast.Declaration,
 	checkConformance checkConformanceFunc,
 ) {
 
+	// process pragmas first, as they determine whether types can later be removed
+	oldRemovedTypes := collectRemovedTypePragmas(validator, oldDeclaration.DeclarationMembers().Pragmas())
+	removedTypes := collectRemovedTypePragmas(validator, newDeclaration.DeclarationMembers().Pragmas())
+
+	// #typeRemoval pragmas cannot be removed, so any that appear in the old program must appear in the new program
+	// they can however, be added, so use the new program's type removals for the purposes of checking the upgrade
+	for oldRemovedType, _ := range oldRemovedTypes { //nolint:maprange
```

Comment:
Sorry, just noticed that this loop has side-effects (reporting), so this needs to be changed.

Maybe switch to an `OrderedSet` to get both deterministic ordering and fast set inclusion tests.